src/fairness/weighted_tree/test/load_accounting_db.cpp

```diff
@@ -0,0 +1,302 @@
+            sqlite3_close (DB);
+
+            return nullptr;
+        }
+
+        // execute SQL statement
+        rc = sqlite3_step (b_select_associations_stmt);
+        while (rc == SQLITE_ROW) {
+            std::string username = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 0));
+            std::string user_shares = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 1));
+            std::string user_job_usage = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_associations_stmt, 3));
+
+            // add user as a child of the node
+            user_node = std::make_shared<weighted_tree_node_t> (
+                            parent_bank,
+                            username,
+                            true,
+                            std::stoll (user_shares),
+                            std::stoll (user_job_usage));
+            node->add_child (user_node);
+            leaf_bank_usage += std::stoi (user_job_usage);
+
+            rc = sqlite3_step (b_select_associations_stmt);
+        }
+        if (rc == SQLITE_ERROR) {
+            fprintf (stderr, "%s\n", sqlite3_errmsg (DB));
+            sqlite3_close (DB);
+
+            return nullptr;
+        }
+        // aggregate usage from user nodes up to their respective banks
+        // and up to the root bank
+        for (std::shared_ptr<weighted_tree_node_t> curr_ancestor = node;
+             curr_ancestor != nullptr;
+             curr_ancestor = curr_ancestor->get_parent ()) {
+
+            curr_ancestor->set_usage (curr_ancestor->get_usage ()
+                                      + leaf_bank_usage);
+        }
+    }
+    // otherwise, this bank has sub banks, so call this
+    // function again with the first sub bank it found
+    else {
+        parent_bank = node;
+        while (rc == SQLITE_ROW) {
+            // execute SQL statement
+            std::string bank = reinterpret_cast<char const *> (
+                sqlite3_column_text (b_select_sub_banks_stmt, 0));
+            banks.push_back (bank);
+            rc = sqlite3_step (b_select_sub_banks_stmt);
+        }
+        if (rc == SQLITE_ERROR) {
+            fprintf (stderr, "%s\n", sqlite3_errmsg (DB));
+            sqlite3_close (DB);
+
+            return nullptr;
+        }
+        for (std::string b : banks) {
+            // reset the prepared statements back to their initial state and
+            // clear their bindings
+            sqlite3_clear_bindings (b_select_associations_stmt);
+            sqlite3_reset (b_select_associations_stmt);
+            sqlite3_clear_bindings (b_select_sub_banks_stmt);
+            sqlite3_reset (b_select_sub_banks_stmt);
+            sqlite3_clear_bindings (b_select_shares_stmt);
+            sqlite3_reset (b_select_shares_stmt);
+            get_sub_banks (DB,
```

Comment:
I think you want to handle an error case here as well? When `get_sub_banks` return `nullptr`?

Reply:
Sorry, I think I might need some explanation here. How does one typically check for `nullptr` when calling a recursive function?

Reply:
I think I can do something like this:

```c++
if (get_sub_banks (DB,
                   b,
                   parent_bank,
                   b_select_shares_stmt,
                   b_select_sub_banks_stmt,
                   b_select_associations_stmt) == nullptr) {
    std::cerr << "get_sub_banks () returned a nullptr" << std::endl;
    return nullptr;
}
```

That way, if the recursive call returns a `nullptr`, an error message gets logged and a `nullptr` gets returned.